### ISSUE TITLE
Adding NOT NULL column with a default is flagged as backwards-incompatible

### ISSUE THREAD

**Issue by grschafer** [3YOURMIND/django-migration-linter#103]
Adding a new column to an existing table seems to trigger an error for the `NOT_NULL` test even if a default is provided. This is contrary to the [docs list of incompatible migrations](https://github.com/3YOURMIND/django-migration-linter/blob/master/docs/incompatibilities.md). Here's example output from `sqlmigrate` using postgresql and django 1.11:

```
BEGIN;
--
-- Add field surname to person
--
ALTER TABLE "contacts_person" ADD COLUMN "surname" varchar(100) DEFAULT '' NOT NULL;
ALTER TABLE "contacts_person" ALTER COLUMN "surname" DROP DEFAULT;
COMMIT;
```

It looks like the `NOT_NULL` test looks for `SET DEFAULT`, but it doesn't look for just `DEFAULT`. As far as I can tell from my searching, `SET DEFAULT` is sql syntax for altering columns and just `DEFAULT` is the sql syntax for creating columns and this syntax is consistent between mysql and postgresql.

Hopefully I'm not missing something silly 😅but if this issue seems legitimate, I'm happy to try out making a PR.

(Thanks for open-sourcing this package!)

**Comment by David-Wobrock** [comment]
Hi @grschafer ! Good call, it seems that https://github.com/3YOURMIND/django-migration-linter/blob/master/docs/incompatibilities.md is incorrect, or at least ambiguous.

As you can see in the SQL that you provided, in Django, when adding a field with a default value is actually:
* creating the field with the default (so fill existing database rows with the default), and
* dropping the default value (because the default is handled by Django, at an application level)

Therefore, an older version of the code which does not know the new column, will see the error: `surname cannot be NULL`, because the database expects the column to have a value provided.

So it is not a bug, but one of the main reasons of this package: to avoid this pitfall.

I'll try updating the doc to be clearer. And I suggest using https://github.com/3YOURMIND/django-add-default-value which allows you to add a database default value to fields that should have one. The linter should work correctly with this package :)

Thanks for your interest in the package! :)

**Comment by grschafer** [comment]
Good point about old code not being able to handle the new non-nullable column in my original example! So, to add a non-nullable field (with a default value) in a backwards-compatible way would take 2 steps:

1. Add field with `null=True` in one migration
1. Set field to `null=False` (i.e. add `NOT NULL` constraint) in next migration

This approach is flagged by `django-migration-linter` as unsafe because django doesn't set the field's default in the database (django itself fills in the default value when creating a model, before sending SQL to the database).

The `django-add-default-value` package you link does add a default at the database-level, so it produces SQL like

```
ALTER TABLE "contacts_person" ALTER COLUMN "surname" SET DEFAULT '';
```

which is detected by this package (`django-migration-linter`) if the default is added in the same migration as the `NOT NULL` constraint is added -- this case is treated as a safe/compatible migration.

Sorry for re-hashing all the above, just thinking through things out loud and maybe it will be helpful to others in the future. I think I'm going to move forward with this approach -- thanks for the helpful response!

I did want to bring up one bit of dissonance with this approach, though. If I'm giving a django field a more complex default (for example, `uuid.uuid4`) then I have to add the `django_add_default_value.AddDefaultValue` migration step with some other static placeholder value since the database doesn't understand `uuid.uuid4`. This might confuse others when they read my code and see 2 different column defaults defined in different places, and they might not know which one takes precedence. This can be cleared up with documentation and comments (which is my current plan!), but I was wondering what a cleaner alternative might be. Perhaps when `django-migration-linter` sees a line of sql like:

```
ALTER TABLE "contacts_person" ALTER COLUMN "surname" SET NOT NULL;
```

it could parse it and see if the django field has a default value set:

```
from django.apps import apps
from django.db.models.fields import NOT_PROVIDED

model = next(m for m in apps.get_models() if m._meta.db_table == 'contacts_person')
has_django_default = model._meta.get_field('surname').default != NOT_PROVIDED
```

and not flag the migration as incompatible if `has_django_default` is true.

But, this approach feels kinda brittle, and I'm not sure if it's safe across different db adapters or django versions. Right now, I'm content with using both packages and adding comments. Thanks again for the help!

**Comment by David-Wobrock** [comment]
> I did want to bring up one bit of dissonance with this approach, though. If I'm giving a django field a more complex default (for example, uuid.uuid4) then I have to add the django_add_default_value.AddDefaultValue migration step with some other static placeholder value since the database doesn't understand uuid.uuid4.

The main reason why you can define any custom function as the default value for a Django field, is that the default value is handled at the application level, by Django itself. You cannot map Python code to DB operations.
Thus, you are only able to add a NOT NULL column in a backward compatible way with a default value that can be expressed with a database operation.

The idea of integrating and having a different approach for "database default" and "django default" is smart and can allow configuring the linter in the way that corresponds to your concerns. I like the idea.
However,
> not flag the migration as incompatible if has_django_default is true.

would kinda undo the problem that the linter is trying to solve
